Create a generic EventCollector

sbrocket opened the pull request:
This creates a new collector impl that can be used to collect arbitrary event types, rather than the specific event types supported by the simpler collectors.

sbrocket committed:
Add RelatedId::contains method

sbrocket committed:
Create EventCollector and wire up to ShardRunner

sbrocket committed:
Add event type filter to EventCollector

sbrocket committed:
Allow filtering events on multiple IDs instead of just one

sbrocket committed:
Update EventCollector filter type to FnMut

This required removing some apparently unused derived Clone impls from
gateway types. The other collectors could be updated similarly.

sbrocket committed:
Add example EventCollector usage to e10_collectors

sbrocket committed:
Create CollectorError, update EventCollectorBuilder future output

sbrocket committed:
Implement EventFilter::validate_options and add tests

sbrocket committed:
Create EventType::related_ids() that returns which ids may exist

This impl is genereated from the same data used to generate the
Event::(user|guild|channel|message)_id methods. This ensures that the
impls always remain consistent; a given field of RelatedIdsForEventType
struct will only ever be false if the corresponding method always
returns Never.

drklee3 commented:
I believe bumping the MSRV would be a breaking change so this and #1430 would have to be retargeted to the `next` branch (no breaking changes on the `current` branch).

sbrocket commented:
I can replace the Iterator::reduce usage with Iterator::fold and avoid bumping the MSRV then. Didn't realize bumping the MSRV was considered a breaking change.

sbrocket committed:
Switch Iterator::reduce to fold to avoid bumping MSRV

sbrocket committed:
Delete accidental newline

sbrocket committed:
Remove dbg! lines inadvertently left in

sbrocket commented:
Any comments or changes needed on this?

Zaitara commented on src/client/bridge/gateway/mod.rs:
```diff
@@ -69,7 +69,7 @@ use crate::gateway::ConnectionStage;
 // Once we can use `Box` as part of a pattern, we will reconsider boxing.
 #[allow(clippy::large_enum_variant)]
-#[derive(Clone, Debug)]
+#[derive(Debug)]
```
Why have you removed it?

sbrocket replied:
This allows `EventFilter` and `event_collector::FilterOptions` to be `!Clone`. It turned out that these types that I removed the `Clone` derive from were never actually cloned, only moved. Removing the `Clone` bound allows the collector's filter closure to be `FnMut` without needing a lock, so that the closure can capture mutable data. Contrast with the existing collectors' respective `FilterOptions` types which use `Arc<dyn Fn<...>>`.

If it turns out that these types ever need to be cloned in the future `FilterOptions` can be updated from `Box<dyn FnMut<...>>` to `Arc<Mutex<FnMut<...>>>`, but for now that's not needed.

Zaitara commented on src/collector/event_collector.rs:
```diff
@@ -0,0 +1,443 @@
+    UnboundedReceiver as Receiver,
+    UnboundedSender as Sender,
+};
+#[cfg(all(feature = "tokio_compat", not(feature = "tokio")))]
```
```suggestion
#[cfg(not(feature = "tokio"))]
```

sbrocket replied:
Can you clarify why you're suggesting this? What I have currently matches the #[cfg] directive in all the existing collectors. Either way, having `tokio` disabled without `tokio_compat` enabled will cause a build failure, either because there's no imported `sleep` and `Sleep` (what I have now) or because this is trying to import something that doesn't exist (your suggestion).

Zaitara replied:
It is to be consistent with the other cfg restriction below. But such an issue should not allowed in the `build.rs` file no?

sbrocket replied:
build.rs only checks for a missing TLS backend, it doesn't check that you have one of either `tokio` or `tokio_compat` enabled.

I'm fine with changing this but I'd rather do it in a separate PR and change all the collectors at the same time if so, and leave this the same as the others for now. How does that sound?

Zaitara commented on src/gateway/mod.rs:
```diff
@@ -157,7 +157,7 @@ impl Display for ConnectionStage {
 /// [`gateway`]: crate::gateway
 /// [`voice`]: crate::model::voice
-#[derive(Clone, Debug)]
+#[derive(Debug)]
```
Same here

drklee3 commented on src/collector/event_collector.rs:
```diff
@@ -0,0 +1,443 @@
+            .field("channel_id", &self.channel_id)
+            .field("guild_id", &self.guild_id)
+            .field("user_id", &self.user_id)
+            .field("user_id", &self.message_id)
```
```suggestion
            .field("message_id", &self.message_id)
```

sbrocket replied:
Fixed

sbrocket committed:
Fix issues with manual FilterOptions Debug impl

sbrocket commented:
Comments addressed

sbrocket commented:
Format CI failed but for unrelated reasons, and it doesn't look like I can trigger a re-run

sbrocket committed:
Switch EventCollector filter to Fn, add Clone back to types

The Fn -> FnMut change will be pushed to next since removing Clone from
the gateway-related types is a breaking change.

sbrocket commented:
Clone removal removed since it's a breaking change and filter closure switched to Fn, will propose updating to FnMut on next.
